File: be/app/schemas/disruption.py
from typing import Literal

from pydantic import Field

from app.schemas.common import ApiModel, LineGeometry, Money, PolygonalGeometry, RiskLevel
from app.schemas.scenario import Facility


class RiskFactor(ApiModel):
    id: str
    label: str


class RoadRisk(ApiModel):
    segment_id: str
    road_name: str
    highway_class: str | None = None
    osm_way_ids: list[str] = Field(default_factory=list)
    geometry: LineGeometry
    risk_probability: float = Field(ge=0, le=1)
    risk_level: RiskLevel
    estimated_delay_minutes: float = Field(ge=0)
    risk_factors: list[RiskFactor]
    affected_supplier_ids: list[str]
    affected_warehouse_ids: list[str]
    affected_order_ids: list[str]


class Route(ApiModel):
    id: str
    type: Literal["baseline", "recovery"]
    origin_facility_id: str
    destination_facility_id: str
    geometry: LineGeometry
    distance_km: float = Field(gt=0)
    eta_minutes: float = Field(gt=0)
    flood_exposure: RiskLevel
    flood_exposure_probability: float = Field(ge=0, le=1)
    affected_road_segment_ids: list[str]


class PrioritizedIssue(ApiModel):
    id: str
    severity: RiskLevel
    subject: str
    description: str


class OperationalImpact(ApiModel):
    impacted_supplier_ids: list[str]
    impacted_warehouse_ids: list[str]
    impacted_order_ids: list[str]
    road_segments_at_risk: int = Field(ge=0)
    sales_exposure: Money
    issues: list[PrioritizedIssue]


class DisruptionAnalysis(ApiModel):
    simulation_id: str
    facilities: list[Facility]
    roads: list[RoadRisk]
    routes: list[Route]
    historical_flood_geometry: PolygonalGeometry | None = None
    impact: OperationalImpact

File: fe/src/domain/disruption.ts
import { z } from "zod";
import { lineGeometrySchema, moneySchema, polygonGeometrySchema, riskLevelSchema } from "./common";
import { facilitySchema } from "./scenario";

export const riskFactorSchema = z.object({ id: z.string(), label: z.string() });
export const roadRiskSchema = z.object({
  segmentId: z.string(), roadName: z.string(),
  highwayClass: z.string().optional(), osmWayIds: z.array(z.string()).default([]),
  geometry: lineGeometrySchema,
  riskProbability: z.number().min(0).max(1), riskLevel: riskLevelSchema,
  estimatedDelayMinutes: z.number().nonnegative(), riskFactors: z.array(riskFactorSchema),
  affectedSupplierIds: z.array(z.string()), affectedWarehouseIds: z.array(z.string()), affectedOrderIds: z.array(z.string()),
});
export const routeSchema = z.object({
  id: z.string(), type: z.enum(["baseline", "recovery"]), originFacilityId: z.string(), destinationFacilityId: z.string(),
  geometry: lineGeometrySchema, distanceKm: z.number().positive(), etaMinutes: z.number().positive(),
  floodExposure: riskLevelSchema, floodExposureProbability: z.number().min(0).max(1), affectedRoadSegmentIds: z.array(z.string()),
});
export const prioritizedIssueSchema = z.object({ id: z.string(), severity: riskLevelSchema, subject: z.string(), description: z.string() });
export const operationalImpactSchema = z.object({
  impactedSupplierIds: z.array(z.string()), impactedWarehouseIds: z.array(z.string()), impactedOrderIds: z.array(z.string()),
  roadSegmentsAtRisk: z.number().int().nonnegative(), salesExposure: moneySchema, issues: z.array(prioritizedIssueSchema),
});
export const disruptionAnalysisSchema = z.object({
  simulationId: z.string(), facilities: z.array(facilitySchema), roads: z.array(roadRiskSchema), routes: z.array(routeSchema),
  historicalFloodGeometry: polygonGeometrySchema.optional(), impact: operationalImpactSchema,
});

export type RiskFactor = z.infer<typeof riskFactorSchema>;
export type RoadRisk = z.infer<typeof roadRiskSchema>;
export type Route = z.infer<typeof routeSchema>;
export type OperationalImpact = z.infer<typeof operationalImpactSchema>;
export type DisruptionAnalysis = z.infer<typeof disruptionAnalysisSchema>;
